### ISSUE TITLE
TrailBase Crate

### ISSUE THREAD

**Issue by JarWarren** [trailbaseio/trailbase#3]
Is there an ETA for a crate? The "TrailBase as a Rust framework" `custom-binary` example imports it locally. Totally fine, but an eventual crate would be very convenient. Even if the API is subject to undocumented breaking changes.

**Comment by ignatz** [comment]
Thanks for filing the first issue :clap: 

> Is there an ETA for a crate?

Not formally. It's certainly not a lot of work and we could publish a crate today. As you've already alluded to, I just wasn't sure about how ready, useful, and stable the APIs are yet. My naive plan was to steer folks more towards SQL and JS/TS APIs, at least in the beginning.

That said, I acknowledge the irony of not having a "TB-as-a-framework" crate for more experienced users, while experienced users are absolutely the most valuable early users.
Long story short, if you wanted to use a "TB-as-a-framework" crate, I'm more than happy to publish one. If this was more of a "nice-to-have" request, I'd probably focus on other things first. Just let me know.

Since I have your attention: I'm curious to get a sense of what you're planning (if you'd like to share). Adding custom axum handlers is certainly the primary use case but maybe there's also something missing from the CLI. Did you have a look at the JS/TS runtime? Any thoughts? Thanks

**Comment by JarWarren** [comment]
Feel completely free to focus your efforts where you think they're most needed! 

Currently planning on a chess.com wannabe, centered around a different historical variant of chess. Initially, gameplay would be asynchronous via REST. Eventually I'd probably want to support realtime matches with a websocket. A secondary goal is to become fluent in a tech that lets me quickly/cheaply spin up projects without the whole Postgres/Redis/etc song and dance.

Not a JS/TS hater, but I'd rather write Rust (or Go). I found TrailBase by [asking around for a framework](https://www.reddit.com/r/rust/comments/1gy35hz/recommended_frameworks_for_a_server_sqlite/). I can do that with a crate or a local clone though. Breaking changes are expected and encouraged. In the end it's as much about learning/growing with the framework as it is about writing the project.

**Comment by JarWarren** [comment]
> If you haven't seen, https://www.youtube.com/watch?v=7VSVfQcaxFY&t=2s is a look at technical look at lichess, which was built by a single person.

Wow! That is impressive. I had no idea it was just him. Glad to see validation for "monolith + satellites" as the narrator called it.

Well I'll definitely end up experimenting in Rust, but don't feel an obligation to alter any of your existing plans on my account haha. The next several weeks will be filled with holidays, so I'll be away from the computer for a while anyways, with family.

When I do get to coding, I'll be happy to be vocal.

**Comment by BuriKizilkaya** [comment]
I understand your opinion to focus on features and then stabilise the API. Some people are also interested in building web applications with the amazing TrailBase as a backend and for example Leptos or HTMX for UI also in Rust. For such use cases a crate would help.

**Comment by ignatz** [comment]
Thanks for the kind words. It undoubtedly would. Things will be a bit slower over the holidays but hope to get to it soon.

If you're itching a lot, you can consider vendoring TB and using local path dependencies: https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-path-dependencies

Thanks for chiming in

**Comment by vootooguru** [comment]
I also hope to do backend development entirely in Rust, which is one of the reasons why I’m not completely satisfied with PocketBase. Currently, I’m using path dependencies to utilize trailbase-core for custom extensions, but there seems to be an issue with stack overflow. I’m not too concerned about whether it's a crate, as path dependencies are sufficient for now. In actual development, if I encounter any issues, I may be able to contribute to the project by fixing bugs or adding features. Right now, I just want to get started quickly.

**Comment by ignatz** [comment]
Quick update: I did publish some of the dependent crates:

* https://crates.io/crates/trailbase-sqlean
* https://crates.io/crates/trailbase-extension
* https://crates.io/crates/trailbase-sqlite

The "core" crate will need a bit more work and testing to make sure that the admin-dash and auth assets (HTML, JS, images, ...) are bundled correctly

**Comment by ignatz** [comment]
Another quick update, there is a https://crates.io/crates/trailbase now. So far the changes are only in the `dev` and not yet the `main` branch since I wanna do a bit more validation first. It was more finicky than expected due to expected crate package structure:

* had to move the UIs to trailbase-core
* had to move proto definitions to trailbase-core
* had to symlink the trailbase ts client into trailbase-core and create a nested pnpm workspace
* had to publish forked packages for refinery. Upstream is rather stale and would have required to downgrade rusqlite (while I'm also experimenting with vendoring rusqlite to get more up-to-date sqlite3 releases)

**Comment by ignatz** [comment]
With the crate released and the relevant changes pushed to head, I'll consider this issue closed.

I'm still reminiscing and am open to input whether there should be any relationship between the major versions of the various trailbase crates and binary releases. My inclination is no, to just let the semantic versions evolve independently and rely on versioning constraints.
